### PR TITLE
mkgmap{,-splitter}: make deterministic add missing phase hook calls

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

This PR makes changes to the derivations of `mkgmap` and `mkgmap-splitter` so that the builds are deterministic.

This was done in 2 parts:
- set `motificationtime` for the `jar` ant task, which removes the impure timestamp from inside the jar file
- create the `.properties` files manually, as otherwise there will be an auto-generated comment at the top of the file with a "last edited" timestamp
  - this also allowed be to get rid of some of the patching being done

The other changes in the PR include adding `runHook {pre,post}{Build,Check,Install}` calls to the respective phases.

There are also some small formatting changes, which I can revert if anyone thinks they are unnecessary.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by TomaSajt** [comment]
> Result of `nixpkgs-review pr 294708` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> <details>
>   <summary>1 package failed to build:</summary>
>   <ul>
>     <li>mkgmap</li>
>   </ul>
> </details>
> <details>
>   <summary>1 package built:</summary>
>   <ul>
>     <li>mkgmap-splitter</li>
>   </ul>
> </details>
> 
> ```
>        > BUILD FAILED
>        > /private/tmp/nix-build-mkgmap-4918.drv-0/mkgmap-r4918/build.xml:302: Test failed.  See test-reports/index.html
> ```

Huh, that's weird... OfBorg seems to have been able to build it without a problem.
I don't have a darwin machine to build this on, so I'd appreciate it if you looked into where exactly it failed.

**Comment by sikmir** [comment]
```xml
<h2>Failures</h2>
<table class="details" border="0" cellpadding="5" cellspacing="2" width="95%">
<tr valign="top">
<th>Name</th><th>Status</th><th width="80%">Type</th><th nowrap>Time(s)</th>
</tr>
<tr valign="top" class="Failure">
<td><a name="testJavaRules"></a>testJavaRules</td><td>Failure</td><td>expected:&lt;0&gt; but was:&lt;1&gt;
<br>
<br>
<code>junit.framework.AssertionFailedError: expected:&lt;0&gt; but was:&lt;1&gt;
<br>
at uk.me.parabola.imgfmt.app.srt.SrtCollatorTest.testJavaRules(SrtCollatorTest.java:128)
<br>
at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
<br>
</code></td><td>0.031</td>
</tr>
</table>
<div class="Properties">
<a href="javascript:displayProperties('uk.me.parabola.imgfmt.app.srt.SrtCollatorTest');">
Properties &raquo;
</a>
</div>
```

**Comment by sikmir** [comment]
> OfBorg seems to have been able to build it without a problem.

That's indeed strange. Looks like build is impure.

**Comment by TomaSajt** [comment]
> > OfBorg seems to have been able to build it without a problem.
> 
> That's indeed strange. Looks like build is impure.

By the way, was this regression introduced in this PR or was it here before?
If you change the original derivation a little bit so that it needs to be rebuilt locally, does it work?

A possible solution might be setting `-Dfile.encoding=UTF8"`

**Comment by TomaSajt** [comment]
It started failing for me as well, I'll look into it, and perhaps disable the test.

**Comment by sikmir** [comment]
> By the way, was this regression introduced in this PR or was it here before?

Not a regression, the same on master. Hydra is happy https://hydra.nixos.org/build/252879712, but local build fails on that test.

**Comment by TomaSajt** [comment]
Okay, I added a `@Ignore` tag onto the failing test, hopefully this helps locally too.

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 294708` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>mkgmap</li>
    <li>mkgmap-splitter</li>
  </ul>
</details>

File: pkgs/applications/misc/mkgmap/splitter/default.nix [hunk 1]
```diff
@@ -74,7 +95,7 @@ stdenv.mkDerivation rec {
     downloadPage = "https://www.mkgmap.org.uk/download/splitter.html";
     sourceProvenance = with sourceTypes; [
       fromSource
-      binaryBytecode  # deps
+      binaryBytecode # deps
     ];
     license = licenses.gpl2Only;
     maintainers = with maintainers; [ sikmir ];
```
Comment (sikmir) [inline pkgs/applications/misc/mkgmap/splitter/default.nix:101]:
```suggestion
    maintainers = with maintainers; [ sikmir ];
    mainProgram = "splitter";
```

**Review by sikmir — APPROVED** [review]

File: pkgs/applications/misc/mkgmap/splitter/default.nix (hunk 1 above)
Comment (TomaSajt) [reply pkgs/applications/misc/mkgmap/splitter/default.nix]: I added `mainProgram` and reordered the attributes asciibetically. If you don't want that, I can revert.

File: pkgs/applications/misc/mkgmap/default.nix [hunk 2]
```diff
@@ -53,36 +63,50 @@ stdenv.mkDerivation rec {
...
-      binaryBytecode  # deps
-    ];
+    homepage = "https://www.mkgmap.org.uk/";
     license = licenses.gpl2Only;
+    mainProgram = "mkgmap";
     maintainers = with maintainers; [ sikmir ];
     platforms = platforms.all;
-    mainProgram = "mkgmap";
   };
+  sourceProvenance = with sourceTypes; [
```
Comment (sikmir) [inline pkgs/applications/misc/mkgmap/default.nix:108]: Misplaced?

Comment (TomaSajt) [reply pkgs/applications/misc/mkgmap/default.nix]: Ah, sorry about that, it's fixed now

**Pull request merged**